xmtp_mls/src/client.rs

```diff
@@ -68,6 +81,99 @@ where
         }
     }
 
+    // TODO: Remove this and figure out the correct lifetimes to allow long lived provider
+    fn mls_provider(&self) -> XmtpOpenMlsProvider {
+        XmtpOpenMlsProvider::new(&self.store)
+    }
+
+    pub async fn register_identity(&self) -> Result<(), ClientError> {
+        // TODO: Mark key package as last_resort in creation
+        let last_resort_kp = self.identity.new_key_package(&self.mls_provider())?;
+        let last_resort_kp_bytes = last_resort_kp.tls_serialize_detached()?;
+
+        self.api_client
+            .register_installation(last_resort_kp_bytes)
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn top_up_key_packages(&self) -> Result<(), ClientError> {
+        let key_packages: Result<Vec<Vec<u8>>, ClientError> = (0..KEY_PACKAGE_TOP_UP_AMOUNT)
+            .map(|_| -> Result<Vec<u8>, ClientError> {
+                let kp = self.identity.new_key_package(&self.mls_provider())?;
+                let kp_bytes = kp.tls_serialize_detached()?;
+
+                Ok(kp_bytes)
+            })
+            .collect();
+
+        self.api_client.upload_key_packages(key_packages?).await?;
+
+        Ok(())
+    }
+
+    async fn get_all_active_installation_ids(
+        &self,
+        wallet_addresses: Vec<String>,
+    ) -> Result<Vec<Vec<u8>>, ClientError> {
+        let update_mapping = self
+            .api_client
+            .get_identity_updates(0, wallet_addresses)
+            .await?;
+
+        let mut installation_ids: Vec<Vec<u8>> = vec![];
+
+        for (_, updates) in update_mapping {
+            let mut tmp: HashSet<Vec<u8>> = HashSet::new();
+            for update in updates {
+                match update {
+                    IdentityUpdate::Invalid => {}
+                    IdentityUpdate::NewInstallation(new_installation) => {
+                        // TODO: Validate credential
+                        tmp.insert(new_installation.installation_id);
+                    }
+                    IdentityUpdate::RevokeInstallation(revoke_installation) => {
+                        tmp.remove(&revoke_installation.installation_id);
+                    }
+                }
+            }
+            installation_ids.extend(tmp);
+        }
+
+        Ok(installation_ids)
+    }
+
+    // Get a flat list of one key package per installation for all the wallet addresses provided.
+    // Revoked installations will be omitted from the list
+    pub async fn get_key_packages_for_wallet_addresses(
+        &self,
+        wallet_addresses: Vec<String>,
+    ) -> Result<Vec<VerifiedKeyPackage>, ClientError> {
+        let installation_ids = self
+            .get_all_active_installation_ids(wallet_addresses)
+            .await?;
+
+        let key_package_results = self
+            .api_client
+            .consume_key_packages(installation_ids)
```

Comment:
Consume key packages downloads one Key Package from the server for each installation_id in the list. The server then marks each key package as having been consumed, so that subsequent requests get a different key package. It's the other side of the coin to `top_up_key_packages`